### PR TITLE
Add new objects in a .custom-group entity of segment

### PR DESCRIPTION
**Pull request by Algorush**
Need to merge this [PR in 3DStreet](https://github.com/3DStreet/3dstreet/pull/510) first to respect elevation level of new adding objects.
This branch was created to test the approach when adding objects from the New Entity panel. Described here: https://github.com/3DStreet/3dstreet-editor/issues/408#issuecomment-2080332617

### CODE REVIEW

**Comment by kfarr** [comment]
Checklist to complete v1:
Checklist

- [x]  Choose which branch we are using for this -- there are 2?
- [x]  Errors while hovering for some objects - triage first
- [ ]  Panel should be hidden behind a isBeta claim flag
- [ ]  Review -- are these the layers that we need? Any missing? Any shouldn't be there?
- [ ] first option selected is blank
- [ ] incorrect scroll behavior - dropdown shouldn't scroll, should always be static at top of modal

LATER:
-  LATER addlayerbutton needs focus state (implemented correctly on addlayerpanel closeButton)
-  LATER Long-term should use a catalog.json format that is shared between core and editor. (Maybe even bring core into editor)
-  LATER Long-term issue -- Hover preview – Is it always creating new entities or is it changing the value of 1 entity? Is it properly disposing of the gltf preview?
-  LATER -- how to create input mechanism for user when placing an entity that has required properties (such as a src url when placing an image)
- need thumbnail preview

**Comment by Algorush** [comment]
> Review -- are these the layers that we need? Any missing? Any shouldn't be there?

Still need to add layers (from Doc):

- intersection (I'm going to implement it as a new component - https://github.com/3DStreet/3dstreet/issues/438),
- Splat,
- Image - to see my work (Screenshot?),
- QR anchor (street-qr-anchor)

**Comment by Algorush** [comment]
> LATER Long-term issue -- Hover preview – Is it always creating new entities or is it changing the value of 1 entity? Is it properly disposing of the gltf preview?

Just for information - there is a special entity for the preview, which is created when the panel is opened and when user hover over the cards, its properties change (mixin, position)

**Comment by Algorush** [comment]
Maybe we should add Segment Layer aslo? It could be just empty sidewalk/road/bike segment. But it's better to first create a segment component with approximate parameters:
type (type from Streetmix) - road, sidewalk, bike path, ground, etc;
elevation level -2, -1, 0, 1, 2;
content (variantString from Streetmix) - for each type user can select variant from the list of variants, in accordance with the structure from segments-variants.js. Or we can in 3DStreet allow user to add any options that roughly correspond to the type of segment.

File: src/components/components/AddLayerPanel/AddLayerPanel.component.jsx [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+    activeCamera.getWorldDirection(direction);
+    cameraFrontVec.copy(activeCamera.position);
+    cameraFrontVec.add(direction.multiplyScalar(5));
+
+    // place preview entity in front of camera in 5 meters
+    preEntity.setAttribute('position', cameraFrontVec);
+  };
+
+  const getParentSegmentEl = (element) => {
+    if (element.className.includes('segment-parent')) {
```
Comment (vincentfretin) [inline src/components/components/AddLayerPanel/AddLayerPanel.component.jsx:231]: It depends of what you want to do here, I don't see segment-parent used anywhere so I'm not sure what is the intended behavior of this code.
If you want to test if the element has the class "segment-parent", it's preferable to do
`element.classList.contains('segment-parent'))`
The line `element.className.includes('segment-parent')` will match any class with segment-parent in it, like `prefix-segment-parent-suffix`.

https://developer.mozilla.org/en-US/docs/Web/API/Element/classList

File: src/components/components/AddLayerPanel/AddLayerPanel.component.jsx [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+  AFRAME.scenes[0].appendChild(preEntity);
+
+  const previewEntity = () => {
+    if (selectedOption == 'Models: Buildings') {
+      preEntity.setAttribute('scale', { x: 0.1, y: 0.1, z: 0.1 });
+    } else {
+      preEntity.setAttribute('scale', { x: 0.5, y: 0.5, z: 0.5 });
+    }
+    /* place object in front of active camera */
+    const direction = new THREE.Vector3();
```
Comment (vincentfretin) [inline src/components/components/AddLayerPanel/AddLayerPanel.component.jsx]: You can move direction variable out of the function so it's not created every time.

File: src/components/components/AddLayerPanel/AddLayerPanel.component.jsx (hunk 1 above)
Comment (Algorush) [reply src/components/components/AddLayerPanel/AddLayerPanel.component.jsx:231]: > The line element.className.includes('segment-parent') will match any class with segment-parent in it, like prefix-segment-parent-suffix.

Segment classes look like this: `segment-parent-1`, `segment-parent-2`, ... So I use `includes` method here.
https://github.com/3DStreet/3dstreet/blob/7017c3db1e3aff5bfe2e0167f12ed1325119984e/src/aframe-streetmix-parsers.js#L797

> It depends of what you want to do here, I don't see segment-parent used anywhere so I'm not sure what is the intended behavior of this code.

This is the code I use to find the parent segment for the selected element. To later get the `data-elevation-posY` attribute from `segment-parent` element, which contains the segment height value. This value is used to position the preview element according to the height of the segment (elevation level)

Comment (vincentfretin) [reply src/components/components/AddLayerPanel/AddLayerPanel.component.jsx:231]: Thanks for the explanation, it makes sense. All good then!

**Comment by Algorush** [comment]
Regarding the preview method - did we settle on the current method of previewing an element - appearing in the middle of a segment? So I can remove the old code related to the preview of the element 5 meters in front of the camera?

**Comment by kfarr** [comment]
@Algorush yes previewing on currently selected segment is good to start

**Comment by Algorush** [comment]
issues to fix:
- delete preview entity when closing the panel or assign an id. Now a new one is created every time the panel is opened
- I see strange duplicates of created elements inside a-scene
- determine the order of adding elements not inside the segment-parent - to buildings-parent, ... There is a small task here on how to do this so that buildings/elements are added to the right place on the scene and inside the scene graph

**Comment by kfarr** [comment]
close by #428

**Pull request closed without merging**